adds live page preview on hover :)

palashkaria opened the pull request:
- Works by adding an iframe to the page on load, & checking for mouseover of page references. If found, changes the iframe src -- this leads to instant preview :)
To see it in action, see the video here:
https://twitter.com/palashkaria/status/1259589041688768512

Draft PR, WIP
- currently works for [[]] links in text. Working on #, & headings etc.
- Need to test this with other roam features.
- Can make the size/styling of the preview customisable? (currently hard-coded)

Stvad commented:
would be nice if it were possible to move mouse over popup and scroll that Iframe/etc

also should block references get preview as-well? (I'm not sure about it myself)

Stvad commented:
hmm, this seems to somehow add a small lag for when I'm typing things in the block (not trying to do hover preview or anything)

Stvad commented:
maybe just because we need to re-render 2 instances of Roam in the same thread?   
wonder if this can be done in a background process or something 🤔

Stvad commented:
if the reference is close to the sight side of the page the popup can go outside of the screen

palashkaria commented:
@Stvad

1. cursor into popover for scroll - this can be a little complex -- as we need to figure out when to close etc. The code itself will be complex

2. popup getting cut off the screen -- I do plan to handle this, but with a minimal impl, but if we want it to be perfect, we might want to look at some small lib, like [tippy](https://atomiks.github.io/tippyjs/). This might help us with the first issue too. Will do some research when implementing.

3. Not sure about the lag -- might be because we have a mouseover handler? I don't think a roam running in an iframe would cause any issues. I also do a dom checks, on mouse over, which might be causing repaints/layout thrashing. Do you have inspect element open when you get the lag? I will try and do a perf profile to see if we can pinpoint the issue.

4. Preview works by loading an iframe -- for block reference, we load the page?
Might not make much sense as the context of the data is lost.

Also, currently the toggle does not work -- did not get to fixing it yet.

Stvad commented:
* 1,2 using a lib sounds good to me
* lag - I don' think it's mouseover - I'm not touching the mouse or anything, just typing the text in the block. Why I think it can be related to Roam in iframe, it's an aggregate of the following:
  - Roam is pretty slow in general. 
  - It does re-rendering when you type things. 
  - Now you have 2 instances of Roam running in your frontend thread, so you need to do 2x amount of live rendering for each typing event
* Block references - no, if you construct URL the way you do right now, but putting the block uid vs page uid you'll get a "zoomed in" view of the block

Not quite sure what you mean by toggle?

palashkaria commented:
@Stvad we are hiding the iframe using opacity/height/width, set to 0; There will be no repaint/relayout happening in that case. We can easily check for this using profiling in chrome's devtools

- AH, you mean Block references + it's children? Could be done, but for longer lengths it will get tricky.

palashkaria commented:
I've been using roam with this enabled for some time now.. Don't see any lag, really. Perf profile also shows 60fps frames

![image](https://user-images.githubusercontent.com/2976363/81512881-d8f03680-9341-11ea-9c01-04eca8bac953.png)

palashkaria commented:
Will keep a lookout for lag -- should see it if it happens; I have this enabled anyway

Stvad commented:
ok, rendering is the wrong word, I suppose, background processing would be more precise.   
When the preview is enabled: I have these moments when the "scripting" consumes 100% CPU (if I read this correctly) and the Roam freezes up. 

![image](https://user-images.githubusercontent.com/81347/81513002-f093d480-92d9-11ea-854f-86e1588b63c4.png)

the following is without preview enabled:

![image](https://user-images.githubusercontent.com/81347/81513046-54b69880-92da-11ea-8057-aa3bdaf12df9.png)

in both cases I'm just typing in the text into the block

Stvad commented:
another thing to support here is Markdown aliases `[alias]([[Page Name]])`

palashkaria commented:
@Stvad could send a dump of the log with preview enabled? It can be done by using rightclick-> save

Stvad commented:
do you mean console log or saving the captured profile?

palashkaria commented:
captured profile @Stvad

Stvad commented:
put it here: https://file.io/o5rQDqPn

palashkaria commented:
@Stvad getting a 404 on that URL

palashkaria commented:
@Stvad I', still not getting the profile you are getting.. Maybe it's the size of the page, etc. Can you please check that link? Maybe something got missed while pasting it here

palashkaria commented:
@Stvad also, I have all other features disabled; only this running. Can you try that too?

Stvad commented:
@palashkaria  hrm, another attempt: https://file.io/xW8m0Gwn

Stvad commented:
oh, it seems those links are 1-time use

Stvad commented:
https://send.firefox.com/download/0f95ee924b341ebf/#fk0CxJsBcqEn73JPxpJ9Lw

palashkaria commented:
Got it, thanks

Also, are you running the other features? or just the live preview?
and what other browser extensions have you got? I'm currently using brave, with minimal extensions & only this feature, I'm not facing this problem as of now.

palashkaria commented:
going by the minified source, I see it's triggered from a 'write' function; which seems to be triggered whenever a new block is created; post this, roam seems to do a server sync. 

That's why you see this randomly -- it happens when the sync is happening, and the main thread gets busy. This will need further debugging (or atleast a clean repro case). I'm running perf profile constantly to see if I get this

Stvad commented:
just fyi - tested in incognito mode, with all the other features disabled - same story.

Stvad commented:
yeah sync makes sense as a culprit - it's run not only when new blocks are created, but when you just add this to existing one, which is what I'm doing

Stvad commented:
a note - my main DB is rather large (19mb unpacked JSON file), just tried this on the smaller DB - it's far less pronounced, still get small stuttering (with corresponding cpu spikes) here and there, but a lot better then with the main db

Stvad commented:
another important thing to confirm is that we're doing the same thing while recording the profiles. As mentioned above - the issues occurs while I'm typing the text into the block (not moving mouse, not creating new blocks or links or anything like it, just continuously enter new text) - are you testing the same scenario?

palashkaria commented:
I just tried in fresh chrome profile, still nothing..  THe worse I get is this; which is still atleast 30fps

![image](https://user-images.githubusercontent.com/2976363/81588224-d63f2100-93d5-11ea-8a71-3c43a5358e86.png)


Could be due to the db size too, maybe? We should ask other people to test it, if possible

palashkaria commented:
Found some other issues in the impl -- meanwhile thinking of how to fix them. For example, one is - if we show the popup, & then later change the underlying link, the popup will never get hidden (as it uses mouseout -- but the element is gone!)

palashkaria commented:
@Stvad yes, I tested out by entering data in the same block - i even see it updating lie in the hovered popup without much perf issue

My db is tiny, just a few kbs -- probably why I'm not facing this.

palashkaria commented:
Some updates
- Switched to using [popper.js](https://popper.js.org/) for positioning, 
- Added support for scrolling into the tooltip and interacting with it (scroll the preview page)
- Remove popper if you edit the item -- and the original item you hovered on is removed from dom

Stvad commented:
shiny new updates!   😃 
Just noticed a bit of weirdness - I have the CSS to enable the scroll bar in Chrome. But when preview is shown it gets hidden for some reason

palashkaria commented:
@Stvad can you share the snippet you are using?

Stvad commented:
```css
::-webkit-scrollbar {
    width: 7px;
}
```

palashkaria commented:
Also, for note:
-  it now supports `#[[page refs]]` :) 
- there are some issues with the astrolobe scolling up as soon as the iframe loads, tried a fix for that, not too sure if it worked

Stvad commented:
btw just had a thought about doing this without the iframe - we can pull the content of the block/page from Datascript. In that case we'd need to do the rendering ourselves though =\, as we'd only get the JSON.

This would probably fix the performance issues, but seems non-trivial

palashkaria commented:
That would be a good 2nd cut; We could leverage some of the code from https://github.com/aravindballa/gatsby-theme-andy 
for parsing/dispalying the content. Would it be possible to use React to do that?

palashkaria commented:
I think we are already injecting React into page/ using react for shortcuts?

Stvad commented:
Yep, we are. That's actually one of the questions I have for this CR - why not use React to abstract away styling/rendering things

Stvad commented:
btw another thing I noticed - if I used preview for the large page - the whole interfaces freezes up. Can we make loading async?

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,173 @@
+
+const createPreviewIframe = () => {
+    const iframe = document.createElement('iframe')
+    const currentDate = new Date()
+    // Tomorrow starts at 2am for the purposes of this
+    currentDate.setHours(currentDate.getHours() - 2)
```
why do you need this here?

palashkaria replied:
this is for the first page that we load by default -- I had kept it as 'today', but in the userscript I made, I kept this the 'search' page. I will do that here too

Stvad replied:
I think you can just use "baseUrl"

palashkaria replied:
@Stvad just the base url is /page, which is not a valid page. Refactored to using daily notes

Stvad replied:
not `basePageUrl` but `baseUrl` which would be something like `https://roamresearch.com/#/app/roam-toolkit` and will lend you on the daily notes

palashkaria commented:
@Stvad if thr loading is async, it'll freeze up anyway. Also, we might never even see the preview -- as the astrolabe will keep spinning.

Stvad commented:
that's sad. I think we can probably ship it with performance as-is (not sure if we can do much about it with the iframe approach?) but disabled by default + with warning that it'd impact performance on the large DBs and that people should use it at their own risk /etc

palashkaria commented:
@Stvad that sounds good! I'l disable by default and add the warning

palashkaria commented:
> Yep, we are. That's actually one of the questions I have for this CR - why not use React to abstract away styling/rendering things

That's because I did not want to add another overhead here, hence just used plain JS. With react, we might need do hover handling a little differently, and in general that will add a lot more code than just vanilla JS

Stvad commented:
>@Stvad that sounds good! I'l disable by default and add the warning
Ok, I'll try to review code tomorrow/day after then :) 

>Do we have a standard way to add warnings?

Nope =\. To have something simple you can probably just have it in the name. But there is no facilities to designate something as experimental/add warnings/etc.

Stvad commented:
Re React: 
>  That's because I did not want to add another overhead here
I don't using React adds much overhead performance-wise, we already do so after all.

>Plus, we might want to rethink the approach of injecting react, as it would mean multiple copies of react running on the page

I imagine we can use one render root for all the things we inject to the page...

>in general that will add a lot more code than just vanilla JS
I have almost no experience with React. So I have a naive question 😅 - why? I thought making it into component would allow us abstract/separate out the rendering part.. (I'm ok with living it in vanilla js if it's the right approach here, just curious)

palashkaria commented:
@Stvad about multiple react roots, check this, on the future of react (React Fire 🔥)
https://github.com/facebook/react/issues/13525


Here it references an issue about how react attaches event listeners to the document -- so potentially if you have 2 copies of react, one of them can eat up the other's events (If `stopPropagtion` is used for eg :) )

https://github.com/facebook/react/issues/2043

Apart from that, if it's two different react versions, we are injecting into the page -- it slows down the first load because it would need to parse React. This is probably happening already, but not very noticeable

palashkaria commented:
Also, about the changes request: I'll get to them this weekend (24-46 hours) :)

palashkaria commented:
adding something like this for warning:

using react-popper-tooltip internally

Also, this works perfectly, in options page, but some issue is there at least in dev mode, when opening the extension. I face the same in while looking at style applied on hover etc. The rendering seems to be stuck sometimes.

![image](https://user-images.githubusercontent.com/2976363/82084245-9d21ec00-9708-11ea-97b0-f2400c0d0e6c.png)

Stvad commented:
Nice! Btw, just noticed that we can still navigate into the `Live Preview` even though it's empty. Should fix it at some point, but it's outside of scope here..

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    iframe.style.borderRadius = '4px'
+    iframe.id = 'roam-toolkit-preview-iframe'
+
+    const styleNode = document.createElement('style')
+    styleNode.innerHTML = `
+        .roam-topbar {
+            display: none !important;
+        }
+        .roam-body-main {
+            top: 0px !important;
+        }
+        #buffer {
+            display: none !important;
+        }
+        iframe {
+            display: none !important;
+        }
+    `
```
extract style creation into separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    if (isAdded(url)) {
+        return
+    }
+    iframe.src = url
+    iframe.style.position = 'absolute'
+    iframe.style.left = '0'
+    iframe.style.top = '0'
+    iframe.style.opacity = '0'
+    iframe.style.pointerEvents = 'none'
+
+    iframe.style.height = '0'
+    iframe.style.width = '0'
+    iframe.style.border = '0'
+    iframe.style.boxShadow = '0 0 4px 5px rgba(0, 0, 0, 0.2)'
+    iframe.style.borderRadius = '4px'
+    iframe.id = 'roam-toolkit-preview-iframe'
```
extract basic iframe setup into a separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    const iframe = document.createElement('iframe')
+
+    const url = Navigation.getPageUrl('search')
+    const isAdded = (pageUrl: string) => !!document.querySelector(`[src="${pageUrl}"]`)
```
I'm not quite sure what this function checks for, can you give it a more descriptive name? also it seems to be duplicated elsewhere

palashkaria replied:
yes, duplication can be removed. This checks if a particular URL is already added to the dom. It is needed because on every hover, our mouseover is getting triggered multiple times, so we might end up adding the iframe multiple times

Stvad replied:
name it something similar to `hasUrlLoaded(url)` maybe ?

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    document.body.appendChild(iframe)
+    const htmlElement = document.querySelector('html')
+    if (htmlElement) {
+        // to reset scroll after adding iframe
```
why do we need this?

palashkaria replied:
some issue where after adding the iframe, the loader moved up.

Roam can fix it by putting a overflow:hidden when loading -- but until they do that, we can use this hack. Causes no harm :)

Stvad replied:
not quite sure what you mean by "loader moved up"?  
Can you extract this into a separate function and add a more detailed explanation of "why" in the comments?

palashkaria replied:
this can happen: @Stvad  
![image](https://user-images.githubusercontent.com/2976363/82117209-5970c600-978c-11ea-95d4-be9a622c0354.png)

adding a comment too

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            display: none !important;
+        }
+    `
+    iframe.onload = (event: Event) => {
+        ;(event.target as HTMLIFrameElement).contentDocument?.body.appendChild(styleNode)
+    }
```
extract into a separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    document.addEventListener('mouseover', (e: Event) => {
+        const target = e.target as HTMLElement
+        const isPageRef = target.classList.contains('rm-page-ref')
+        const isPageRefTag = target.classList.contains('rm-page-ref-tag')
+        // remove '#' for page tags
+        const text = isPageRefTag ? target.innerText.slice(1) : target.innerText
```
extract into a separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    return iframe
+}
+const enableLivePreview = () => {
+    let hoveredElement: HTMLElement | null
+    let popupTimeout: ReturnType<typeof setTimeout> | null
+    let popper: Instance | null = null
```
this seems like it should be a Class

Stvad replied:
by this I mean the whole larger function

palashkaria replied:
which function? the whole code you mean? Seems like a huge refactor at this point. @Stvad

palashkaria replied:
@Stvad instead of addressing each comment, I just refactored into a class. take a look.

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+        const isPageRefTag = target.classList.contains('rm-page-ref-tag')
+        // remove '#' for page tags
+        const text = isPageRefTag ? target.innerText.slice(1) : target.innerText
+        if (isPageRef) {
```
all the internals should be in a separate function (though they are themselves should be split into separate functions :))

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+                        previewIframe.style.opacity = '1'
+                        previewIframe.style.pointerEvents = 'all'
+
+                        popper = createPopper(target, previewIframe, {
```
popper setup should be in a separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            const isAdded = (pageUrl: string) => !!document.querySelector(`[src="${pageUrl}"]`)
+            const isVisible = (pageUrl: string) =>
+                (document.querySelector(`[src="${pageUrl}"]`) as HTMLElement)?.style.opacity === '1'
+            if ((!isAdded(url) || !isVisible(url)) && previewIframe) {
```
the condition should be extracted into the variable/function with meaningful name

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            const isVisible = (pageUrl: string) =>
+                (document.querySelector(`[src="${pageUrl}"]`) as HTMLElement)?.style.opacity === '1'
+            if ((!isAdded(url) || !isVisible(url)) && previewIframe) {
+                previewIframe.src = url
+                previewIframe.style.height = '500px'
+                previewIframe.style.width = '500px'
+                previewIframe.style.pointerEvents = 'none'
```
whatever this is doing should be extracted into the function that has name that explain what it's doing

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            if (!popupTimeout) {
+                popupTimeout = window.setTimeout(() => {
+                    if (previewIframe) {
+                        previewIframe.style.opacity = '1'
+                        previewIframe.style.pointerEvents = 'all'
```
separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+        const target = e.target as HTMLElement
+        const relatedTarget = e.relatedTarget as HTMLElement
+        const iframe = document.getElementById('roam-toolkit-preview-iframe') as HTMLIFrameElement
+        if (
+            (hoveredElement === target && relatedTarget !== iframe) ||
+            (target === iframe && relatedTarget !== hoveredElement) ||
+            !document.body.contains(hoveredElement)
```
uh, this needs to be a function with the meaningful name. It's subexpressions need to be variables

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            (hoveredElement === target && relatedTarget !== iframe) ||
+            (target === iframe && relatedTarget !== hoveredElement) ||
+            !document.body.contains(hoveredElement)
+        ) {
```
all the internals should be in a separate function (though they are themselves should be split into separate functions :))

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+                iframe.style.height = '0'
+                iframe.style.width = '0'
+            }
+            if (popper) {
```
separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+                        scrollContainer.scrollTop = 0
+                    }
+                }
+                iframe.style.pointerEvents = 'none'
+                iframe.style.opacity = '0'
+                iframe.style.height = '0'
+                iframe.style.width = '0'
```
separate function

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+            clearTimeout(popupTimeout as ReturnType<typeof setTimeout>)
+            popupTimeout = null
+            if (iframe) {
+                if (iframe.contentDocument) {
+                    // scroll to top when removed
+                    const scrollContainer = iframe.contentDocument.querySelector('.roam-center > div')
+                    if (scrollContainer) {
+                        scrollContainer.scrollTop = 0
+                    }
+                }
```
separate function

palashkaria commented on src/ts/components/Tooltip.tsx:
```diff
@@ -0,0 +1,48 @@
+                </StyledTooltipContent>
+            )}
+        >
+            {({getTriggerProps, triggerRef}) => (
```
this is render props -- the children are function

It's from the sample codehere:
https://www.npmjs.com/package/react-popper-tooltip

bascially this passes on mouse to the dom element on which tooltip is to applied. (it is needed by TooltipTrigger)

we can organize a little better, but these are all internal components. Externally, we just need to pass `children` and a `tooltip`

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,169 @@
+    settings: [],
+}
+
+Settings.isActive('live_preview').then(active => {
+    if (active) {
+        enableLivePreview()
+    }
+})
+
+browser.runtime.onMessage.addListener(async message => {
+    if (message?.featureId === 'live_preview' && message?.type === 'toggle') {
+        if (message.value) {
+            enableLivePreview()
+        } else {
+            removeLivePreview()
+        }
+    }
+})
```
@Stvad how to know when settings are updated? I thought we need to use onMessage for that?

Stvad replied:
yeah, but it's structurally simpler, so you just need to add another thing to trigger here https://github.com/roam-unofficial/roam-toolkit/blob/master/src/ts/contentScripts/dispatcher/index.ts#L18

palashkaria replied:
I don't seem to be able to get that working -- the actions in that dispatcher do not seem to be triggered from anywhere. All keyboard shortcuts, for eg, are triggered from their own `onPress`. 

Also I see that `updateShortcuts` will be called on every toggle/update, which might not be needed for live preview.

palashkaria replied:
I can add a seperate trigger for this,  but it doesn't seem right --- we will need to couple all the code for preview to dispatcher -- even now, the dispatcher has a lot of coupling, which makes it difficult to modify

palashkaria replied:
using `settings-updated` for now -- it seems like something needs to change here.
`settings-updated` does not tell us which setting is updated, so it just calls `updateShortcuts` even any setting is toggled. 

I am checking for live-preview setting when any setting is updated, and taking actions accordingly.

Also related: CSS toggle does not really work (even in the published extension). This is because on any toggle, we only dispatch `settings-updated`,which does not impact the CSS.

I have ideas around how to fix this flow to be a little decouples from each other -- that way settings don't know about each other, and it would fix some of these issues. Let's keep them for a separate change @Stvad

Stvad replied:
agree for making settings changes elsewhere.  
For `settings-updated` - it's true that it's coarse (i.e. does not tell you what setting changed etc).  
For the use cases we see I think it's ok though - it's simple. And even though it may trigger reloading of things more often on settings change - that's fine because settings change is a very rare event. 

Replied to you on slack for things in dispatcher not being triggered and did some cleanup

Stvad replied:
for better interface for dispatcher, I imagine we can invert the flow of the dependencies, and make it implement observer. So it'd be an object/class that would have interface like

```javascript
onSettingsUpdated(function)
onSettingUpdated(name, function) //when we'd need a more granular things

...etc
```

palashkaria replied:
Yes, this would be a good pattern too. Similar to onMessage, but less explicit

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+
+let iframeInstance: PreviewIframe | null = null
+
+const setupIframe = (active: boolean) => {
```
this name is misleading (i.e. it does not reflect that it'll destroy the preview if called). `togglePreview` maybe?

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+    private resetIframeForNextHover(iframe: HTMLIFrameElement) {
+        if (iframe) {
+            if (iframe.contentDocument) {
+                // scroll to top when removed, so the next popup is not scrolled
```
should this be using `scrollToTopHack` ?

palashkaria replied:
this is a different scollToTop hack :)

Stvad replied:
why do they have to be different. would scrolling to the top of html not be enough here too?

palashkaria replied:
@Stvad that is during loading. When a post is loaded, you don't scroll on the `html`, you scroll on a div inside `.roam-center`. Need to set the scrolltop of that.

Stvad commented on src/ts/roam/navigation.ts:
```diff
@@ -3,15 +3,29 @@ import {addDays, isValid, RoamDate} from '../date/common'
+        return this.getPageUrl(page[':block/uid'])
+    },
+    getPageUrl(uid?: string) {
+        if (!uid) {
+            return this.baseUrl().toString()
+        }
```
what's use case for this?

palashkaria replied:
to get the daily notes page

Stvad replied:
you can just call `baseUrl` explicitly..

palashkaria replied:
that does not return a string -- don't want to leak .toString() into app code. And we might need the baseUrl as is for some other purpose

Stvad replied:
not sure what do you mean by `leak .toString`. you literally use this only to get baseUrl, and there is already function for that 😛  
I don't find it intuitive that that this function would silently "succeed" on falsey input =\

palashkaria replied:
I use this to convert eh baseUrl (which is the class URL) to a string. It's not the same thing. @Stvad Should I add a separate function for this? What should it be called? 

By leak `toString`, I mean exactly that. It needs to be in a function here.

palashkaria replied:
added a function `getDailyNotesUrl` for this @Stvad

Stvad replied:
hrm. ok 😛

Stvad commented on src/ts/utils/settings.ts:
```diff
@@ -41,9 +43,15 @@ export const Settings = {
 export const prepareSettings = (features: Feature[]): Feature[] => {
-    return features.map((feature: any) => {
+    return features.map((feature: Feature) => {
         if (feature.toggleable !== false) {
             feature.toggleable = true
         }
```
in the prev comment about this - I meant more like extracting this, alongside with the similar code you had for `enabledByDefault` to set up the default values on the feature object

palashkaria replied:
this is different from that -- this causes side effects by mutating the feature, that does not. let's not mix it up. @Stvad

Stvad replied:
it was before :ton

palashkaria replied:
yes :) fixed it

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    private removeIframe() {
+        if (this.iframe && document.body.contains(this.iframe)) {
```
If we try to remove an iframe which doesn't exist in `body`, there will be an exception. So both of these need to be true at the same time @Stvad the condition is correct

Stvad replied:
I meant having something like:
```javascript
if(document.body.contains(this.iframe)) { //false for undefined/null/etc
    document.body.removeChild(this.iframe)
}
if (this.iframe) this.iframe = null
```

palashkaria replied:
@Stvad typescript will throw an error there

palashkaria replied:
@Stvad will need to check for `this.iframe` first. Updated

Stvad replied:
hrm, it's now more complicated then before 🙈.   
is typescript error because the `this.iframe` can still potentially be null?

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+    private getIFrameByUrl(url: string): HTMLIFrameElement | null {
+        return document.querySelector(`iframe[src="${url}"]`)
+    }
+    private getIsIFrameVisibleByUrl(url: string): boolean {
```
@Stvad this is about being 'visible', not being in preview -- visible is if opacity is set to `1`. Maybe `isUrlPreviewVisible`

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+        this.attachMouseOutListener(iframe)
+    }
+
+    private attachMouseOverListener(iframe: HTMLIFrameElement) {
```
@Stvad this is passed before we add the iframe to the instance; not sure of the implication of changing this

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+    private attachMouseOutListener(iframe: HTMLIFrameElement) {
+        document.addEventListener('mouseout', (e: MouseEvent) => {
+            const target = e.target as HTMLElement
+            const relatedTarget = e.relatedTarget as HTMLElement
```
yes. That's the name on `event`

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+    activate() {
+        this.iframe = this.initPreviewIframe()
+    }
+    destroy() {
```
popper is cleaned up on mouse out -- but we can potentially have a case when no mouse out happens & someone disables it. Would be safer to clean up

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,293 @@
+
+class PreviewIframe {
+    iframeId = 'roam-toolkit-iframe-preview'
+    iframe: HTMLIFrameElement | null = null
```
how about constructing this (and maybe calling activate in general?) as a part of class constructor, so we can can be sure that it's not null and avoid null-checking it everywhere?

Stvad replied:
as after all, iframe should exist during the whole lifecycle of this class

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,293 @@
+        if (!this.iframe) {
+            return
+        }
+        const visibleIframe = this.getVisibleIframeByUrl(url)
+        if (visibleIframe) {
+            // if visible, just return the iframe
+            return
+        }
```
should use class iframe instance

palashkaria replied:
@Stvad will not work -- mouse events get called 3 times

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,279 @@
+}
+checkSettingsAndSetupIframe()
+
+browser.runtime.onMessage.addListener(async message => {
```
👀

palashkaria replied:
@Stvad doing this causes the iframe removal to stop working, because of the same instance problem -- on toggle, it reloads this script, resulting in the instance being `null`. 

If used via onMessage, it works fine. 


Changes I made (which caused it to break
- in live-preview, remove this `browser.runtime.onMessage.addListener` handling. 

- in `dispatcher`

```
 ['settings-updated', () => {
        updateShortcuts()
        checkSettingsAndSetupIframeToggle()
    }
```

- now on toggle, it does not `.destroy()` as it does not find the instance.

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,293 @@
+
+    private getIFrameByUrl(url: string): HTMLIFrameElement | null {
+        return document.querySelector(`iframe[src="${url}"]`)
+    }
```
@Stvad Haha, I wish 😅 It is not possible in prettier because of the way prettier works.See this more more info
> It turns out that empty lines are very hard to automatically generate. The approach that Prettier takes is to preserve empty lines the way they were in the original source code
https://prettier.io/docs/en/rationale.html#empty-lines

Stvad replied:
huh, interesting. May I suggest you configure your IDE to do so then ?😛

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,293 @@
+
+    private initPreviewIframe() {
+        const url = Navigation.getPageUrl()
+        const existingIframe = this.getIFrameByUrl(url)
```
No clear answer. Can happen, have seen mouse events getting triggered multiple times.

palashkaria commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    private getIFrameByUrl(url: string): HTMLIFrameElement | null {
+        return document.querySelector(`iframe[src="${url}"]`)
+    }
```
As mentioned, not sure. We get mouse events triggering multiple times, and these are to make sure that we don't add the iframes multiple times. Will not be able to debug and figure out immediately @Stvad

palashkaria commented:
@Stvad as discussed, we have the problem of the script getting executed multiple times, and so multiple instances being created. Added a check based on ID for now, and simplified some more things in https://github.com/roam-unofficial/roam-toolkit/pull/52/commits/9afd28e272ff75e329d28824352c3631aa56f72d

palashkaria commented:
might have found the root cause for this issue; we are loading the same contentscript multiple times -- once directly, once in dispatcher, for eg. that's why when I move the code to dispatcher, it causes problems. Each content script is supposed to be isolated on it's own. See this for more details: 
https://stackoverflow.com/questions/26240463/how-do-i-re-use-code-between-content-scripts-in-a-chrome-extension

> @Stvad as discussed, we have the problem of the script getting executed multiple times, and so multiple instances being created. Added a check based on ID for now, and simplified some more things in [9afd28e](https://github.com/roam-unofficial/roam-toolkit/commit/9afd28e272ff75e329d28824352c3631aa56f72d)

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -61,40 +62,34 @@ class PreviewIframe {
-        const existingIframe = this.getIFrameByUrl(url)
+        const url = Navigation.getDailyNotesUrl()
+        const existingIframe = this.getExisitingIframe()
         if (existingIframe) {
```
I hope we'd be getting rid of this now. but if we were not  - I think the assignment would have been still required, to ensure that we operate on the appropriate underlying object

palashkaria replied:
this can be removed after #58 gets merged. I tested by merging with #57 already :)

Stvad commented on src/ts/contentScripts/live-preview/index.tsx:
```diff
@@ -0,0 +1,267 @@
+        this.iframe.style.width = '0'
+    }
+
+    private scrollToTopOnMouseOut() {
```
nit: can you place the scroll to top hacks close to each other and extract commonality between them?

Stvad commented on src/ts/features/livePreview.tsx:
```diff
@@ -25,12 +25,16 @@ browser.runtime.onMessage.addListener(async message => {
-    } else if (iframeInstance) {
-        iframeInstance.destroy()
-        iframeInstance = null
+    if (active) {
```
why this?

Stvad replied:
i.e. why change from previous version of this
